docs/source/notes/autograd.rst

```diff
@@ -58,6 +58,122 @@ will also require them.
+
+Enable no-grad mode when you need to perform operations that should not be
+tracked by autograd, but you’d still like to use the outputs of these
+computations in grad mode later.
+
+No-grad mode should be used when you have a block of code or a function
+for which no gradient computation is needed, or when you need to perform
+an in-place update on a parameter that should not be tracked by autograd.
+
+For example, no-grad mode might be useful when writing an optimizer. When
+performing the training update you’d like to update parameters
+in-place without the update being tracked by autograd.
+You also intend to use the updated parameters for computations in
+grad mode in the next forward pass.
+
+The implementations in the `nn.init <nn-init-doc_>`_ package also
+rely on no-grad mode when initializing the parameters as to avoid
+autograd tracking when updating the intialized parameters in-place.
+
+Inference Mode
+^^^^^^^^^^^^^^
+
+Inference mode is an extreme version of no-grad mode. Just like in no-grad
+mode, computations in inference mode are not tracked by autograd, but
+enabling inference mode will allow PyTorch to speed up your model even more.
+This better runtime comes with a drawback: tensors created in inference mode
+will not be able to be used in computations tracked by autograd after
+exiting inference mode.
+
+Enable inference mode when you are performing computations that don’t need
+tracked for the gradient computation, AND you don’t plan on using the tensors
+created in inference mode in any computation that is tracked by autograd later.
+
+It is recommended that you try out inference mode in the parts of your code
+that do not require autograd tracking. If it works out of the box
+for your use case it’s a free performance win. If you run into errors after
+enabling inference mode, check that you are not using tensors created in
+inference mode outside inference mode. It may also be possible that inference
+mode does not apply to your use case, in which case you can always switch
+back to no-grad mode.
+
+For implementation details of inference mode please see
+`RFC-0011-InferenceMode <https://github.com/pytorch/rfcs/pull/17>`_.
+
+Evaluation Mode (``nn.Module.eval()``)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Evaluation mode is not actually a mechanism to locally disable gradient computation.
+It is included here anyway because it is sometimes confused to be such a mechanism.
+
+Functionally, ``module.eval()`` (or equivalently ``module.train()``) are completely
+orthogonal to no-grad mode and inference mode. How ``model.eval()`` affects
+your model depends entirely on the specific modules used in your model and
+whether they define any training-mode specific behavior.
+
+You are responsible for calling ``model.eval()`` and ``model.train()`` if your
+model relies on modules such as :class:`torch.nn.Dropout` and
+:class:`torch.nn.BatchNorm` that may behave
```

Comment:
looks like `BatchNorm` isn't actually a proper class in nn. Is there some canonical batch norm doc to link to or we'd have to link to something like `BatchNorm1d`.

Reply:
BatchNorm2d should be a good generic example.